mat64/cholesky.go

```diff
@@ -178,3 +178,162 @@ func (s *SymDense) InverseCholesky(chol *Cholesky) error {
 	s.SymOuterK(1, &t)
 	return err
 }
+
+// SymRankOne performs a rank-1 update of the original matrix A and refactorizes
+// its Cholesky factorization, storing the result into the reciever. That is, if
+// in the original Cholesky factorization
+//  U^T * U = A,
+// in the updated factorization
+//  U'^T * U' = A + alpha * x * x^T = A'.
+//
+// Note that when alpha is negative, the updating problem may be ill-conditioned
+// and the results may be inaccurate, or the updated matrix A' may not be
+// positive definite and not have a Cholesky factorization. SymRankOne returns
+// whether the updated matrix A' is positive definite.
+//
+// SymRankOne updates a Cholesky factorization in O(n²) time. The Cholesky
+// factorization computation from scratch is O(n³).
+func (c *Cholesky) SymRankOne(orig *Cholesky, alpha float64, x *Vector) (ok bool) {
+	n := orig.Size()
+	if x.Len() != n {
+		panic(matrix.ErrShape)
+	}
+	if orig != c {
+		if c.isZero() {
+			c.chol = NewTriDense(n, matrix.Upper, nil)
+		} else if c.chol.mat.N != n {
+			panic(matrix.ErrShape)
+		}
+		c.chol.Copy(orig.chol)
+	}
+
+	if alpha == 0 {
+		return true
```

Comment:
This is incorrect at the moment. When Factorize returns false, the original matrix is not positive definite while here by returning true we pretend it is. Again it boils down to how Cholesky should behave when Factorize returns false. Should we add a bool field to Cholesky indicating whether the matrix is PD and return it here? Or can we rely on cond being infinity? Should we panic and document it? The solve routines should behave similarly (now they return an error based on cond _after_ performing the solve).